nifi-nar-bundles/nifi-standard-services/nifi-distributed-cache-services-bundle/nifi-distributed-cache-client-service/src/main/java/org/apache/nifi/distributed/cache/client/NettyHandshakeHandler.java

```diff
@@ -0,0 +1,150 @@
+        logger.info("Magic header written");
+        final int currentVersion = versionNegotiator.getVersion();
+        final ByteBuf byteBufVersion = Unpooled.wrappedBuffer(new OutboundAdapter().write(currentVersion).toBytes());
+        ctx.writeAndFlush(byteBufVersion);
+        logger.info("Protocol version {} proposed", versionNegotiator.getVersion());
+    }
+
+    @Override
+    public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+        if (promiseHandshakeComplete.isSuccess()) {
+            ctx.fireChannelRead(msg);
+        } else {
+            final ByteBuf byteBuf = (ByteBuf) msg;
+            try {
+                processHandshake(ctx, byteBuf);
+            } catch (IOException | HandshakeException e) {
+                throw new IllegalStateException(e);
+            } finally {
+                byteBuf.release();
+            }
+        }
+    }
+
+    /**
+     * Negotiate distributed cache protocol version with remote service.
+     *
+     * @param ctx     the {@link Channel} context
+     * @param byteBuf the byte stream received from the remote service
+     * @throws HandshakeException on failure to negotiate protocol version
+     * @throws IOException        on write failure
+     */
+    private void processHandshake(final ChannelHandlerContext ctx, final ByteBuf byteBuf) throws HandshakeException, IOException {
+        final short statusCode = byteBuf.readUnsignedByte();
+        if (statusCode == ProtocolHandshake.RESOURCE_OK) {
+            logger.info("Protocol version {} accepted", versionNegotiator.getVersion());
+            protocol.set(versionNegotiator.getVersion());
+        } else if (statusCode == ProtocolHandshake.DIFFERENT_RESOURCE_VERSION) {
+            final int newVersion = byteBuf.readInt();
+            logger.info("Protocol version {} counter proposal", newVersion);
+            final Integer newPreference = versionNegotiator.getPreferredVersion(newVersion);
+            Optional.ofNullable(newPreference).orElseThrow(() -> new HandshakeException("Could not agree on protocol version"));
+            versionNegotiator.setVersion(newPreference);
+            ctx.writeAndFlush(Unpooled.wrappedBuffer(new OutboundAdapter().write(newPreference).toBytes()));
+        } else if (statusCode == ProtocolHandshake.ABORT) {
+            final short length = byteBuf.readShort();
+            final byte[] message = new byte[length];
+            byteBuf.readBytes(message);
+            throw new HandshakeException("Remote destination aborted connection with message: " + new String(message, StandardCharsets.UTF_8));
+        } else {
+            throw new HandshakeException("Unknown handshake signal: " + statusCode);
+        }
+    }
+
+    @Override
+    public void channelReadComplete(final ChannelHandlerContext ctx) {
+        if (promiseHandshakeComplete.isSuccess()) {
+            ctx.fireChannelReadComplete();
+        } else if (protocol.get() > 0) {
```

Comment:
Recommend using a named static variable instead of `0` to indicate what is being evaluated.